### DISCUSSION TITLE
Proposal: Generic software hosting configuration.

### DISCUSSION THREAD

**Discussion by nielsbasjes** [renovatebot/renovate#10604] [Suggest an Idea]
I have this proposal for an improvement:

For the retrieval of the changelog/release notes of an updated dependency 
- from github.com the GITHUB_COM_TOKEN is needed. https://docs.renovatebot.com/self-hosting/#githubcom-token-for-release-notes
- from same system of where renovate is running the RENOVATE_TOKEN is used.

This works up to the point when there are more code hosting systems from where the release notes are to be retrieved.

So if I pull a dependency that is hosted on gitlab.com or something else (like many of the apache tools that are hosted on for example https://gitbox.apache.org/repos/asf?p=commons-lang.git ) I cannot get the changelog at all.

So I have this proposal to improve this:

`Make it generically possible to have an additional list of "sourcecode hosting platforms for dependencies" in the config.`

Looking at what I already see I would expect something like this (rough sketch):

```json
  "hostRules": [
    {
      "matchHost": "github.com",
      "hostType": "github",
      "endPoint": "https://api.github.com/",
      "token": "{{ secrets.GITHUB_COM_TOKEN }}"
    },
    {
      "matchHost": "git.example.com",
      "hostType": "gitlab",
      "endPoint": "https://git.example.com/api/v4/",
      "token": "{{ secrets.GITLAB_EXAMPLE_COM_TOKEN }}"
    },
    {
      "matchHost": "gitlab.com",
      "hostType": "gitlab",
      "endPoint": "https://gitlab.com/api/v4/",
      "token": "{{ secrets.GITLAB_COM_TOKEN }}"
    }
  ]
```

NOTES: 
- With this people can add multiple (internal/specific) code hosting systems depending on their needs.
- Doing it in this way the github.com solution can be provided in backward compatible way by providing a standard config upstream.
- I think the token/username+password should be `recommended but optional` in all cases.
- I suspect that looking at the `gitbox.apache.org` examples that a kind of templating will be needed to extract the actual file from the hosting system.
- Looking through the release notes code I spotted very specific things like `baseUrl.includes('gitlab')` and `if (baseUrl !== 'https://gitlab.com/')` which I suspect should change.

Is this a good idea?

**Comment by rarkins** [comment]
We should definitely make our changelog/release notes fetching more generic and less tied to github/gitlab, and any step towards that is a good idea.

> So if I pull a dependency that is hosted on gitlab.com or something else (like many of the apache tools that are hosted on for example [gitbox.apache.org/repos/asf?p=commons-lang.git](https://gitbox.apache.org/repos/asf?p=commons-lang.git) ) I cannot get the changelog at all.

That should not be the case - gitlab is supported. But maybe it was a typo because the URL you pointed to doesn't appear to be gitlab.

But even in this example it would appear that Renovate would need platform-awareness to understand that page.

I think it should work like this:

- Changelog function receives sourceUrl and/or changelogUrl
- Function determines if that sourceUrl is a host type which we have *special* handling for (requires awareness of the platform/host type, because their APIs are different)
- If we don't have custom handling for the platform type then we fall back to trying to do a regular HTTP-based file fetch and parsing

I don't understand all of what you're proposing but what I think we need is:
- Refactor the current changelog so it's less "if github else if gitlab" and instead ready for extended platforms
- Add any new platforms we need (e.g. bitbucket, azure devops, gitea.. and whatever that apache.org gitbox is)
- Make sure it's clear to users how to configure if it doesn't work out of the box

For platforms which don't require authentication then of course we shouldn't require authentication. But for e.g. GitHub I'm not willing to budge on that because their API will quickly block your IP if requests are unauthenticated and it's a very poor user experience when that happens - better to fetch none at all than to have random success.

**Reply by nielsbasjes** [reply]
Correct: That URL is not gitlab but it is the url posted in the pom.xml of apache commons-lang.
https://repo1.maven.org/maven2/org/apache/commons/commons-lang3/3.12.0/commons-lang3-3.12.0.pom

This tool is the "official" webinterface for git: "gitweb" 

https://git-scm.com/book/en/v2/Git-on-the-Server-GitWeb
https://git-scm.com/docs/gitweb

And this thing (as far as I know) does not have authentication.

I fully understand the github situation, so I'm cool with that. 

For the rest of your reply: Sounds awesome!
